### PR TITLE
8318200: String::newStringNoRepl and String::getBytesNoRepl does not throw CharacterCodingException

### PR DESCRIPTION
**Pull request by wenshao**
When calling String::newStringNoRepl and String::getBytesNoRepl, we need to use try...catch to handle CharacterCodingException and throw IllegalArgumentException instead of CharacterCodingException to make the calling code cleaner.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8318200](https://bugs.openjdk.org/browse/JDK-8318200): String::newStringNoRepl and String::getBytesNoRepl does not throw CharacterCodingException (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16209/head:pull/16209` \
`$ git checkout pull/16209`

Update a local copy of the PR: \
`$ git checkout pull/16209` \
`$ git pull https://git.openjdk.org/jdk.git pull/16209/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16209`

View PR using the GUI difftool: \
`$ git pr show -t 16209`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16209.diff">https://git.openjdk.org/jdk/pull/16209.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16209#issuecomment-1765375772)

### CODE REVIEW

**Review by liach — CHANGES_REQUESTED** [review]
The thrown CCE was intended for `Files.readString` and `Files.writeString`, and is in reality, only used by those 2 sites. However, you missed updating those sites, so they now throw IAE instead of CCE, which should be fixed.

**Comment by cl4es** [comment]
I agree with @liach that `Files.readString / writeString` should continue to throw `CCE`, not `IAE`.

**Comment by AlanBateman** [comment]
> The thrown CCE was intended for `Files.readString` and `Files.writeString`, and is in reality, only used by those 2 sites. However, you missed updating those sites, so they now throw IAE instead of CCE, which should be fixed.

Right, these APIs are specified to throw an IOException "If an I/O error occurs reading from the file or a malformed or unmappable byte sequence is read". Changing this to throw an unexpected IAE would be a breaking change. So good to spot this. Need to make sure that tier2 tests are run with this change.

File: test/jdk/java/lang/String/NoReplTest.java [hunk 1]
```diff
@@ -59,8 +59,12 @@ public void newStringNoReplTest() throws IOException {
                             .withPrefix("x")
                             .withUpperCase()
                             .formatHex(read.getBytes(UTF_16)));
-        } catch (CharacterCodingException cce) {
```
Comment (cl4es) [inline test/jdk/java/lang/String/NoReplTest.java:62]: Shouldn't these test changes be completely reverted now?

File: src/java.base/share/classes/java/nio/file/Files.java [hunk 2]
```diff
@@ -3351,7 +3352,14 @@ public static String readString(Path path, Charset cs) throws IOException {
...
         if (path.getClass().getModule() != Object.class.getModule())
             ba = ba.clone();
-        return JLA.newStringNoRepl(ba, cs);
+        try {
+            return JLA.newStringNoRepl(ba, cs);
+        } catch (IllegalArgumentException e) {
+            if (e.getCause() instanceof CharacterCodingException cause) {
+                throw cause;
+            }
+            throw e;
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/file/Files.java:3361]: This is wrong, if IAE is thrown for any other reason than CharacterCodingException then it means the JDK is broken.

**Comment by AlanBateman** [comment]
I've set this PR to require two Reviewers as there is obvious breakage here and no evidence that the tier2 tests have been run.

**Comment by rgiulietti** [comment]
@wenshao The proposal changes the exception types for seemingly no reason.
What concrete problem does this enhancement PR address?

**Comment by wenshao** [comment]
> @wenshao The proposal changes the exception types for seemingly no reason. What concrete problem does this enhancement PR address?

@lianch 's suggestion in PR #14812 : 
"Make the NoRepl methods throw IllegalArgument (like NoRepl1 do) by default; move the exception handling logic into Files implementation, the only place where it's actually needed. "

We used noRepl in UUID/HexFormat to handle CCE. We felt that the code was not clean enough, so we submitted this PR.

After the discussion above, I found that things are not that simple. Should I continue this PR?

**Review by RogerRiggs — CHANGES_REQUESTED** [review]
Hacking around the exception types isn't a good fix.
If the exception is thrown, it is because of a charset conversion error.
IllegalArgumentException is for cases where the arguments are incorrect.

**Comment by AlanBateman** [comment]
> The new usages that are driving this change would be better served by a method dedicated to creating latin1 strings from a caller provided byte array. The JavaLangAccess shared secret interface is already overused, but it is cleaner, more maintainable, and fit for purpose to add a method than to contort the exception handling as proposed.

We have to be very careful with proposals like this as it means code outside of java.lang having access to the underlying bytes. I think other alternatives needs to be explored to avoid this and related concerns.

**Comment by RogerRiggs** [comment]
> We have to be very careful with proposals like this as it means code outside of java.lang having access to the underlying bytes. I think other alternatives needs to be explored to avoid this and related concerns.

Yes, adding another method in the "NoRepl" family needs to be done with care and be narrowly focused.
JLA already exposes several methods that allow the contents of strings to be read from the underlying value or created from caller provided buffers. All to avoid extra allocations and copying.

**Comment by Glavo** [comment]
Should we revisit https://github.com/openjdk/jdk/pull/14655?

**Pull request closed without merging**